src/DependencyGraph/DependencyGraphHelpers.js

```diff
@@ -8,24 +8,89 @@
  */
 'use strict';
 
+const fs = require('fs');
 const path = require('fast-path');
 
 class DependencyGraphHelpers {
-  constructor({ providesModuleNodeModules, assetExts }) {
-    this._providesModuleNodeModules = providesModuleNodeModules;
+  constructor({ roots, providesModuleNodeModules, assetExts }) {
+    this._roots = roots;
+    this._providesModuleNodeModuleDirs = this._resolveHastePackages(providesModuleNodeModules);
     this._assetExts = assetExts;
   }
 
+  /**
+   * An item in the `providesModuleNodeModule` array is an object containing two keys:
+   *  * "name" - the package name
+   *  * "parent" - if a string, it's the name of the parent package of the module in
+   *  the dep tree. If null, it signifies that we should look for this package at
+   *  the top level of the module tree.
+   *
+   * We try to resolve the specified module...but if we can't, we fallback to looking
+   * in "node_modules/[package name]" (this satisfies use cases where the actual
+   * "react-native" package is one of the root dirs, such as when we run tests
+   * or examples).
+   */
+  _resolveHastePackages(packages) {
+    const packagePathForPackage = ({ name, parent }, rootDir) => {
+      let packagePath;
+      if (parent) {
+        if (!Array.isArray(parent)) {
+          parent = [parent];
+        }
+        parent.push(name);
+        packagePath = path.join(
+          rootDir,
+          'node_modules',
+          parent.join(path.sep + 'node_modules' + path.sep)
+        );
+      } else {
+        packagePath = path.join(rootDir, 'node_modules', name);
+      }
+
+      if (packagePath[packagePath.length - 1] === path.sep) {
```

Comment:
This can happen in the case of a providesModuleNodeModule entry that looks like:

```{ name: 'some-module', parent: ['parentPackage1', 'parentPackage2'] }```, as the path will end with `/node_modules/`